### PR TITLE
Reporting IDE errors

### LINKED ISSUE #359: enabling the IDEA's error/exception reporting mechanism
"the plugin developers haven't implemented the feature" (reporting errors from the IDE error monitoring and tracking (an example: https://github.com/Aalto-LeTech/intellij-error-reporting )

### PR DESCRIPTION
**Pull request by Taikelenn**
# Description of the PR
Closes #359.

IntelliJ's functionality of capturing exception stack traces has been implemented. The remaining thing to add is a POST request with the JSON payload, and obviously some backend server (probably not part of this repository)

~TODO: include platform information (operating system, IntelliJ version, plugin version, other plugins installed)~

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: reports-server/config.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+SERVER_ADDRESS = ""
+SERVER_PORT = 25
+SERVER_LOGIN = ""
+SERVER_PASSWORD = ""
+
+FROM_ADDRESS = ""
+TO_ADDRESS = ""
+MAIL_SUBJECT = "A+ Courses error logs"
```
Comment (OlliKiljunen) [inline reports-server/config.py:8]: This is excellent!

File: reports-server/main.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
+from http.server import ThreadingHTTPServer
```
Comment (OlliKiljunen) [inline reports-server/main.py:1]: On the [documentation](https://docs.python.org/3/library/http.server.html), it says:
> Warning: http.server is not recommended for production. It only implements basic security checks.

Have you taken that warning into account?

File: reports-server/reportsubmitter.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            logging.info(f"Report file {report_file} sent")
+
+def signal_exit():
+    exit_event.set()
+
+def report_submitter_thread():
+    logging.info("Starting the report submitter thread")
+    while True:
+        _send_pending_reports()
+        if exit_event.wait(5): # sleep for 15 minutes
```
Comment (OlliKiljunen) [inline reports-server/reportsubmitter.py:54]: Should `5` be changed?

File: reports-server/reportsubmitter.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    if not awaiting_reports:
+        return
+
+    for report_file in awaiting_reports:
+        report_path = os.path.join(REPORTS_PATH, report_file)
+        with open(report_path, "r", encoding="utf-8") as report:
+            mail_content_string += f"<<< Report file {report_file} >>>\n{report.read().strip()}\n<<< End of report file >>>\n\n"
+
+    mail_content_string = f"This is a report containing {len(awaiting_reports)} report{'s' if not len(awaiting_reports) == 1 else ''} from the A+ Courses plugin.\n\n{mail_content_string}"
```
Comment (OlliKiljunen) [inline reports-server/reportsubmitter.py:40]: Nice plurals ;)

**Review by OlliKiljunen — APPROVED** [review]
Great! 👍 Looks like a very good implementation. Please check my comments, and that those are alright. Then feel free to merge.

File: reports-server/reportsubmitter.py (hunk 3 above)
Comment (Taikelenn) [reply reports-server/reportsubmitter.py:54]: Yes

File: reports-server/main.py (hunk 2 above)
Comment (Taikelenn) [reply reports-server/main.py:1]: The server itself will not be exposed to the internet, it will be behind a reverse proxy (I will setup `nginx`)

**Review by stellatsv — APPROVED** [review]

**Pull request merged**